Mathlib/GroupTheory/Submonoid/Units.lean

```diff
@@ -0,0 +1,322 @@
+
+@[to_additive (attr := simp)]
+lemma units_inf (S T : Submonoid M): (S ⊓ T).units = S.units ⊓ T.units :=
+ofUnits_units_gc.u_inf
+
+@[to_additive]
+lemma units_sInf {s: Set (Submonoid M)} : (sInf s).units = ⨅ S ∈ s, S.units :=
+ofUnits_units_gc.u_sInf
+
+@[to_additive]
+lemma units_iInf {ι : Sort*} {f : ι → Submonoid M} : (iInf f).units = ⨅ (i : ι), (f i).units :=
+ofUnits_units_gc.u_iInf
+
+@[to_additive]
+lemma units_iInf₂ {ι : Sort*} {κ : ι → Sort*} {f : (i : ι) → κ i → Submonoid M} :
+    (⨅ (i : ι), ⨅ (j : κ i), f i j).units = ⨅ (i : ι), ⨅ (j : κ i), (f i j).units :=
+    ofUnits_units_gc.u_iInf₂
+
+@[to_additive (attr := simp)]
+lemma units_bot : (⊥ : Submonoid M).units = ⊥ := ofUnits_units_gci.u_bot
+
+@[to_additive]
+lemma units_surjective : Function.Surjective (Submonoid.units (M := M)) :=
+ofUnits_units_gci.u_surjective
+
+@[to_additive]
+lemma units_left_inverse :
+    Function.LeftInverse (Submonoid.units (M := M)) (Subgroup.ofUnits (M := M)) :=
+    ofUnits_units_gci.u_l_leftInverse
+
+/-- The multiplicative equivalence between the type of units of `M` and the submonoid of unit
+elements of `M`. -/
+@[to_additive (attr := simps!) " The additive equivalence between the type of additive units of `M`
+  and the additive submonoid whose elements are the additive units of `M`. "]
+noncomputable def unitsTypeEquivIsUnitSubmonoid :
+  Mˣ ≃* IsUnit.submonoid M where
+  toFun x := ⟨x, Units.isUnit x⟩
+  invFun x := x.prop.unit
+  left_inv x := IsUnit.unit_of_val_units _
+  right_inv x := by simp_rw [IsUnit.unit_spec]
+  map_mul' x y := by simp_rw [Units.val_mul]; rfl
+
+/-- The equivalence between the subgroup of units of `S` and the submonoid of unit
+elements of `S`. -/
+@[to_additive (attr := simps!) " The equivalence between the additive subgroup of additive units of
+`S` and the additive submonoid of additive unit elements of `S`.  "]
+noncomputable def unitsEquivIsUnitSubmonoid (S : Submonoid M) : S.units ≃* IsUnit.submonoid S :=
+S.unitsEquivUnitsType.trans unitsTypeEquivIsUnitSubmonoid
+
+/-- The unit group of a unit group is equivalent to the same group. -/
+@[to_additive " The additive unit group of an additive unit group is equivalent to the same
+group. " ]
+def unitsTypeUnitsTypeEquivUnitsType : Mˣˣ ≃* Mˣ := toUnits.symm
+
+end Units
+
+end Submonoid
+
+namespace Subgroup
+
+@[to_additive (attr := simp)]
+lemma mem_ofUnits_iff (S : Subgroup Mˣ) (x : M) : x ∈ S.ofUnits ↔ ∃ y ∈ S, y = x := Iff.rfl
+
+@[to_additive]
+lemma mem_ofUnits (S : Subgroup Mˣ) {x : M} {y : Mˣ} (h₁ : y ∈ S) (h₂ : y = x) : x ∈ S.ofUnits :=
+  ⟨_, h₁, h₂⟩
+
+@[to_additive]
+lemma exists_unit_coe_eq_mem_ofUnits (S : Subgroup Mˣ) {x : M} (h : x ∈ S.ofUnits) :
+    ∃ y ∈ S, y = x := h
+
+@[to_additive]
+lemma mem_of_mem_coe_ofUnits (S : Subgroup Mˣ) {y : Mˣ} : (y : M) ∈ S.ofUnits → y ∈ S :=
+  fun ⟨_, hm, he⟩ => (Units.ext he) ▸ hm
+
+@[to_additive]
+lemma isUnit_of_mem_ofUnits (S : Subgroup Mˣ) {x : M} : x ∈ S.ofUnits → IsUnit x :=
```

Comment:
Same here.

Reply:
As in, make `x` explicit?

Reply:
Oh I see.

Reply:
Same as previous, resolve there if necessary.